{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction to Biology with Python \n",
    "\n",
    "> Adapted from Fatih Enes Kemal Ergin (https://notebook.community/eneskemalergin/OldBlog/_oldnotebooks/Biology_with_Python)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Overview\n",
    "\n",
    "In this tutorial, we will give a brief introduction to biological concepts with theory and implementation in Python.\n",
    "\n",
    "This tutorial is designed for programmers with some basic knowledge of biology. If you're new to biology, we recommend checking out the [Best Resources To Learn Molecular Biology For A Computer Scientist](https://www.biostars.org/p/3066/) before proceeding. Those with a strong background in biology can skip directly to the implementation sections.\n",
    "\n",
    "## Learning Objectives\n",
    "\n",
    "- Understand basic biological concepts such as DNA, RNA, and proteins\n",
    "- Learn about transcription and translation processes\n",
    "- Implement DNA to protein translation in Python\n",
    "- Estimate molecular weight of biological sequences\n",
    "\n",
    "## Prerequisites\n",
    "\n",
    "- Basic understanding of Python programming\n",
    "- Familiarity with Jupyter notebooks\n",
    "\n",
    "## Get Started\n",
    "\n",
    "We will use Python to perform the following tasks:\n",
    "\n",
    "- Translate DNA sequences into protein sequences\n",
    "- Estimate molecular weight of DNA, RNA, and protein sequences\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DNA, RNA, Protein\n",
    "\n",
    "### DNA \n",
    "\n",
    "DNA, or deoxyribonucleic acid, is the fundamental building block of life. It contains the information cells need to synthesize proteins and replicate themselves. In short, DNA serves as the storage repository for the information required for any cell to function.\n",
    "\n",
    "DNA is composed of four nucleotide bases: Adenine (A), Guanine (G), Thymine (T), and Cytosine (C). A DNA sequence typically looks like this: \"ATTGCTGAAGGTGCGG.\"\n",
    "\n",
    "The length of DNA is measured by the number of base pairs it contains, usually expressed in kilobase pairs (kBp) or megabase pairs (mBp). Each base has a complementary base, meaning in the double-helical structure of DNA, Adenine (A) pairs with Thymine (T), and Guanine (G) pairs with Cytosine (C).\n",
    "\n",
    "DNA is organized into tightly wound structures called chromosomes. Humans have 23 pairs of **chromosomes**, which are further divided into smaller segments known as **genes**.\n",
    "\n",
    "Determining the arrangement of the four nucleotide bases in DNA is called **DNA sequencing**. There are various methods for sequencing DNA, often performed using specialized machines or through a process called gel electrophoresis.\n",
    "\n",
    "### RNA\n",
    "\n",
    "RNA is somewhat similar to DNA; both are nucleic acids composed of nitrogen-containing bases linked by a sugar-phosphate backbone. However, structural and functional differences distinguish RNA from DNA.\n",
    "\n",
    "Structurally, RNA is single-stranded, whereas DNA is double-stranded. DNA contains the base Thymine (T), while RNA contains Uracil (U) instead. Additionally, RNA nucleotides include the sugar ribose, unlike DNA, which contains deoxyribose.\n",
    "\n",
    "Functionally, DNA serves as the repository for protein-encoding information, while RNA uses this information to facilitate the synthesis of specific proteins within the cell."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "### Proteins\n",
    "\n",
    "Proteins are essential to all life forms. In fact, it is often said that without proteins, life as we know it would not exist.\n",
    "\n",
    "Proteins are sequences composed of 20 different amino acids, linked together by polypeptide bonds. The structure of a protein directly influences its chemical activity within an organism. Most proteins adopt a three-dimensional (3D) structure, which enables them to perform a wide variety of functions in living organisms. The complexity of protein structure makes it a challenging yet fascinating subject for research.\n",
    "\n",
    "In computer science, proteins are represented as sequences of 20 distinct letters, with each letter corresponding to a specific amino acid. Amino acids linked together in a protein chain are often referred to as residues. The term \"residue\" is commonly used when referring to a specific amino acid at a particular position within a protein chain."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Transcription\n",
    "\n",
    "The process of reading DNA and creating RNA from it is called *transcription*. On the computational side, we will use the same DNA sequence and replace the DNA nucleotide \"T\" with \"U\". The Python code `dna.replace(\"T\", \"U\")` accomplishes this.\n",
    "\n",
    "- The DNA strand that has the same sequence as the RNA is called the **coding strand**.\n",
    "\n",
    "### Translation\n",
    "\n",
    "Most RNA molecules go on to specify protein amino acid sequences in a process called **translation**; these are known as messenger RNAs (mRNA).\n",
    "\n",
    "- Each subsequent group of *three* bases is called a **codon**.\n",
    "- The regions of an RNA chain that are removed are called **introns**, while those that remain are called **exons**.\n",
    "- The presence of introns makes it significantly more challenging to determine which parts of a gene are actually used to produce protein sequences.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Using biological sequences in computing\n",
    "\n",
    "### Important tips\n",
    "\n",
    "- Fit the sequences into appropriate data structures, ensuring they are reusable.\n",
    "- The simplest and most common approach is to store sequences as text:\n",
    "    - DNA and RNA can be represented as strings composed of 4 distinct characters:\n",
    "        - DNA: A, C, G, T\n",
    "        - RNA: A, C, G, U\n",
    "    - Proteins can be represented as strings composed of 20 distinct characters, each corresponding to a specific amino acid.\n",
    "\n",
    "\n",
    "### 1. Translating a DNA sequence into protein\n",
    "\n",
    "In this example, we will translate a given DNA sequence into a protein using pre-defined structures for amino acid (aa) representation. For now, we will focus solely on implementing the core concept of translation, without considering starting codons, stop codons, or special codons."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Here is the genetic code of the amino acids defined as dictionaries\n",
    "STANDARD_GENETIC_CODE = {\n",
    "    \"UUU\": \"Phe\",  # Phenylalanine\n",
    "    \"UUC\": \"Phe\",  # Phenylalanine\n",
    "    \"UCU\": \"Ser\",  # Serine\n",
    "    \"UCC\": \"Ser\",  # Serine\n",
    "    \"UAU\": \"Tyr\",  # Tyrosine\n",
    "    \"UAC\": \"Tyr\",  # Tyrosine\n",
    "    \"UGU\": \"Cys\",  # Cysteine\n",
    "    \"UGC\": \"Cys\",  # Cysteine\n",
    "    \"UUA\": \"Leu\",  # Leucine\n",
    "    \"UCA\": \"Ser\",  # Serine\n",
    "    \"UAA\": None,   # Stop codon\n",
    "    \"UGA\": None,   # Stop codon\n",
    "    \"UUG\": \"Leu\",  # Leucine\n",
    "    \"UCG\": \"Ser\",  # Serine\n",
    "    \"UAG\": None,   # Stop codon\n",
    "    \"UGG\": \"Trp\",  # Tryptophan\n",
    "    \"CUU\": \"Leu\",  # Leucine\n",
    "    \"CUC\": \"Leu\",  # Leucine\n",
    "    \"CCU\": \"Pro\",  # Proline\n",
    "    \"CCC\": \"Pro\",  # Proline\n",
    "    \"CAU\": \"His\",  # Histidine\n",
    "    \"CAC\": \"His\",  # Histidine\n",
    "    \"CGU\": \"Arg\",  # Arginine\n",
    "    \"CGC\": \"Arg\",  # Arginine\n",
    "    \"CUA\": \"Leu\",  # Leucine\n",
    "    \"CUG\": \"Leu\",  # Leucine\n",
    "    \"CCA\": \"Pro\",  # Proline\n",
    "    \"CCG\": \"Pro\",  # Proline\n",
    "    \"CAA\": \"Gln\",  # Glutamine\n",
    "    \"CAG\": \"Gln\",  # Glutamine\n",
    "    \"CGA\": \"Arg\",  # Arginine\n",
    "    \"CGG\": \"Arg\",  # Arginine\n",
    "    \"AUU\": \"Ile\",  # Isoleucine\n",
    "    \"AUC\": \"Ile\",  # Isoleucine\n",
    "    \"ACU\": \"Thr\",  # Threonine\n",
    "    \"ACC\": \"Thr\",  # Threonine\n",
    "    \"AAU\": \"Asn\",  # Asparagine\n",
    "    \"AAC\": \"Asn\",  # Asparagine\n",
    "    \"AGU\": \"Ser\",  # Serine\n",
    "    \"AGC\": \"Ser\",  # Serine\n",
    "    \"AUA\": \"Ile\",  # Isoleucine\n",
    "    \"ACA\": \"Thr\",  # Threonine\n",
    "    \"AAA\": \"Lys\",  # Lysine\n",
    "    \"AGA\": \"Arg\",  # Arginine\n",
    "    \"AUG\": \"Met\",  # Methionine (Start codon)\n",
    "    \"ACG\": \"Thr\",  # Threonine\n",
    "    \"AAG\": \"Lys\",  # Lysine\n",
    "    \"AGG\": \"Arg\",  # Arginine\n",
    "    \"GUU\": \"Val\",  # Valine\n",
    "    \"GUC\": \"Val\",  # Valine\n",
    "    \"GCU\": \"Ala\",  # Alanine\n",
    "    \"GCC\": \"Ala\",  # Alanine\n",
    "    \"GAU\": \"Asp\",  # Aspartic acid\n",
    "    \"GAC\": \"Asp\",  # Aspartic acid\n",
    "    \"GGU\": \"Gly\",  # Glycine\n",
    "    \"GGC\": \"Gly\",  # Glycine\n",
    "    \"GUA\": \"Val\",  # Valine\n",
    "    \"GUG\": \"Val\",  # Valine\n",
    "    \"GCA\": \"Ala\",  # Alanine\n",
    "    \"GCG\": \"Ala\",  # Alanine\n",
    "    \"GAA\": \"Glu\",  # Glutamic acid\n",
    "    \"GAG\": \"Glu\",  # Glutamic acid\n",
    "    \"GGA\": \"Gly\",  # Glycine\n",
    "    \"GGG\": \"Gly\",  # Glycine\n",
    "}\n",
    "\n",
    "# Pre-defined DNA sequence, We will use this along the way.\n",
    "dna_seq = \"ATGGTGCATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGGCAAGGTG\"\n",
    "print(\"Input DNA: \")\n",
    "print(dna_seq)\n",
    "\n",
    "print(\"Output RNA:\")\n",
    "# Convert the DNA sequence to RNA by replacing 'T' with 'U'\n",
    "# This is because RNA uses uracil (U) instead of thymine (T)\n",
    "rna_seq = dna_seq.replace(\"T\", \"U\")\n",
    "print(rna_seq)\n",
    "\n",
    "\n",
    "def protein_translation(seq, genetic_code):\n",
    "    \"\"\"Translate a nucleic acid sequence into a protein sequence.\n",
    "\n",
    "    Translation proceeds until the end of the sequence or until a stop codon is reached.\n",
    "\n",
    "    Args:\n",
    "        seq (str): The nucleic acid sequence to translate.\n",
    "        genetic_code (dict): The genetic code dictionary.\n",
    "\n",
    "    Returns:\n",
    "        list: The translated protein sequence as a list of amino acids.\n",
    "\n",
    "    \"\"\"\n",
    "    # Changes all the T into U, DNA to RNA\n",
    "    seq = seq.replace(\"T\", \"U\")  # Make sure we have RNA sequence\n",
    "    protein_seq = []  # Initializing the protein_seq list to store the output\n",
    "\n",
    "    i = 0\n",
    "    while i + 2 < len(seq):\n",
    "        # Get codons of three letters\n",
    "        codon = seq[i : i + 3]\n",
    "        # Get the match-up amino acid from the genetic code\n",
    "        amino_acid = genetic_code.get(codon)  # Use .get to avoid KeyError for invalid codons\n",
    "        # If found stop looping (stop codon reached)\n",
    "        if amino_acid is None:  # Found stop codon (e.g., UAA, UAG, or UGA)\n",
    "            break\n",
    "        # Otherwise, add the amino acid to protein_seq list\n",
    "        protein_seq.append(amino_acid)\n",
    "        i += 3  # Move to the next codon\n",
    "\n",
    "    return protein_seq\n",
    "\n",
    "\n",
    "print(\"Output Protein:\")\n",
    "# Translate the RNA sequence into a protein sequence using the genetic code\n",
    "print(protein_translation(dna_seq, STANDARD_GENETIC_CODE))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. Estimating the Molecular Weight\n",
    "\n",
    "This example estimates the molecular weight of a DNA, RNA, or protein molecule in units of Daltons. Note that this is only an approximation, as residues can reversibly bind hydrogen ions under varying conditions (e.g., pH affects the binding of Hâº ions to acidic and basic sites). Additionally, the calculation assumes standard isotopic proportions.\n",
    "\n",
    "#### Steps\n",
    "\n",
    "- Define a function with two arguments: sequence and MoleculeType. \n",
    "- Inside the function, create a dictionary to store the average molecular weights of the different residues.\n",
    "- Initialize a variable to accumulate the total molecular mass."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define function with seq and mol_type Protein\n",
    "def estimate_mol_mass(seq, mol_type=\"protein\"):\n",
    "    \"\"\"Calculate the molecular weight of a biological sequence.\n",
    "\n",
    "    Assumes normal isotopic ratios and protonation/modification states.\n",
    "\n",
    "    Args:\n",
    "        seq (str): The biological sequence to calculate the molecular weight for.\n",
    "        mol_type (str, optional): The type of molecule, either \"DNA\", \"RNA\", or\n",
    "          \"protein\". Defaults to \"protein\".\n",
    "\n",
    "    Returns:\n",
    "        float: The molecular weight of the biological sequence in daltons.\n",
    "\n",
    "    \"\"\"\n",
    "    # Define a dictionary with the molecular masses for DNA, RNA, and protein residues\n",
    "    residue_masses = {\n",
    "        \"DNA\": {\"G\": 329.21, \"C\": 289.18, \"A\": 323.21, \"T\": 304.19},  # Masses of DNA nucleotides\n",
    "        \"RNA\": {\"G\": 345.21, \"C\": 305.18, \"A\": 329.21, \"U\": 302.16},  # Masses of RNA nucleotides\n",
    "        \"protein\": {  # Masses of amino acids (protein residues)\n",
    "            \"A\": 71.07,\n",
    "            \"R\": 156.18,\n",
    "            \"N\": 114.08,\n",
    "            \"D\": 115.08,\n",
    "            \"C\": 103.10,\n",
    "            \"Q\": 128.13,\n",
    "            \"E\": 129.11,\n",
    "            \"G\": 57.05,\n",
    "            \"H\": 137.14,\n",
    "            \"I\": 113.15,\n",
    "            \"L\": 113.15,\n",
    "            \"K\": 128.17,\n",
    "            \"M\": 131.19,\n",
    "            \"F\": 147.17,\n",
    "            \"P\": 97.11,\n",
    "            \"S\": 87.07,\n",
    "            \"T\": 101.10,\n",
    "            \"W\": 186.20,\n",
    "            \"Y\": 163.17,\n",
    "            \"V\": 99.13,\n",
    "        },\n",
    "    }\n",
    "\n",
    "    # Get the corresponding mass dictionary based on the specified molecule type (DNA, RNA, or protein)\n",
    "    mass_dict = residue_masses[mol_type]\n",
    "    # Start with the mass of the extra end atoms (H + OH)\n",
    "    mol_mass = 18.02\n",
    "\n",
    "    # Loop through each letter in the biological sequence (DNA, RNA, or protein)\n",
    "    for letter in seq:\n",
    "        # Add the mass of the current residue to the total molecular mass\n",
    "        # If the residue is not found in the dictionary, add 0.0 (for unknown or invalid residues)\n",
    "        mol_mass += mass_dict.get(letter, 0.0)\n",
    "\n",
    "    # Return the calculated molecular mass\n",
    "    return mol_mass\n",
    "\n",
    "\n",
    "# Test Case 1: Calculate the molecular weight of a DNA sequence\n",
    "print(\"DNA: \" + dna_seq)\n",
    "# Call the function with the dna_seq variable and \"DNA\" as the molecule type\n",
    "print(\"Weight: \", estimate_mol_mass(dna_seq, mol_type=\"DNA\"), \" daltons\")\n",
    "\n",
    "# Test Case 2: Calculate the molecular weight of an RNA sequence\n",
    "print(\"RNA: \" + rna_seq)\n",
    "# Call the function with the rna_seq variable and \"RNA\" as the molecule type\n",
    "print(\"Weight: \", estimate_mol_mass(rna_seq, mol_type=\"RNA\"), \" daltons\")\n",
    "\n",
    "# Test Case 3: Calculate the molecular weight of a protein sequence\n",
    "# Define a sample protein sequence\n",
    "protein_seq = \"IRTNGTHMQPLLKLMKFQKFLLELFTLQKRKPEKGYNLPIISLNQ\"\n",
    "print(\"Protein: \" + protein_seq)\n",
    "# Call the function with the protein_seq variable and the default value \"protein\"\n",
    "print(\"Weight: \", estimate_mol_mass(protein_seq), \" daltons\")\n",
    "print()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Conclusion\n",
    "In this module, we've learned about understanding basic biological concepts such as DNA, RNA, and proteins, We also learned about transcription and translation processes, implemented DNA to protein translation in Python, abd estimated molecular weight of biological sequences.\n",
    "\n",
    "## Clean up\n",
    "Remember to shut down your Jupyter Notebook instance when you're done to avoid unnecessary charges. You can do this by stopping the notebook instance from the Amazon SageMaker console."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (S2_T2)",
   "language": "python",
   "name": "s2_t2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
