{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction to Biology with Python \n",
    "\n",
    "> Adapted from Fatih Enes Kemal Ergin (https://notebook.community/eneskemalergin/OldBlog/_oldnotebooks/Biology_with_Python)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Overview\n",
    "\n",
    "In this tutorial, we will give a brief introduction to biological concepts with theory and implementation in Python.\n",
    "\n",
    "This tutorial is designed for programmers with some basic knowledge of biology. If you\"re new to biology, we recommend checking out the [Best Resources To Learn Molecular Biology For A Computer Scientist](https://www.biostars.org/p/3066/) before proceeding. Those with a strong background in biology can skip directly to the implementation sections.\n",
    "\n",
    "## Learning Objectives\n",
    "\n",
    "- Understand basic biological concepts such as DNA, RNA, and proteins\n",
    "- Learn about transcription and translation processes\n",
    "- Implement DNA to protein translation in Python\n",
    "- Estimate molecular weight of biological sequences\n",
    "\n",
    "## Prerequisites\n",
    "\n",
    "- Basic understanding of Python programming\n",
    "- Familiarity with Jupyter notebooks\n",
    "\n",
    "## Get Started\n",
    "\n",
    "We will use Python to perform the following tasks:\n",
    "\n",
    "- Translate DNA sequences into protein sequences\n",
    "- Estimate molecular weight of DNA, RNA, and protein sequences\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DNA, RNA, Protein\n",
    "\n",
    "### DNA \n",
    "\n",
    "DNA or other wise called deoxyribonucleic acid is the building block of the life. It contains the information the cell requires to synthesize protein and to replicate itself, to be short it is the storage repository for the information that is required for any cell to function.\n",
    "\n",
    "There are basically four nucleotide bases, which make up the DNA. Adenine (A), Guanine (G), Thymine (T) and Cytosine(C). A DNA sequence looks some thing like this \"ATTGCTGAAGGTGCGG\". \n",
    "\n",
    "DNA is measured according to the number of base pairs it consists of, usually in kBp or mBp (Kilo/Mega base pairs). Each base has its complementary base, which means in the double helical structure of DNA, A will have T as its complimentary and similarly G will have C.\n",
    "\n",
    "The DNA is broken down into bits and is tightly wound into coils, which are called **chromosomes**; human beings have 23 pairs of chromosomes. These chromosomes are further broken down into smaller pieces of code called **Genes**. \n",
    "\n",
    "DNA is made up of four nucleotide bases, finding out the arrangement of the bases is called **DNA sequencing**, there are various methods for sequencing a DNA, it is usually carried out by a machine or by running the DNA sample over a gel otherwise called gel electrophoresis.\n",
    "\n",
    "### RNA\n",
    "\n",
    "RNA is somewhat similar to DNA; they both are nucleic acids of nitrogen-containing bases joined by sugar-phosphate backbone. However structural and functional differences distinguish RNA from DNA. \n",
    "\n",
    "Structurally, RNA is a single-stranded where as DNA is double stranded. DNA has Thymine (T), where as RNA has Uracil (U). RNA nucleotides include sugar ribose, rather than the Deoxyribose that is part of DNA. \n",
    "\n",
    "Functionally, DNA maintains the protein-encoding information, whereas RNA uses the information to enable the cell to synthesize the particular protein.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "### Proteins\n",
    "\n",
    "Proteins are the most commonly used in life forms, we may say that without proteins we would not exist.\n",
    "\n",
    "They are sequences made up with 20 different amino acids by _polypeptides_. Structure of the protein effects the chemical activity of the body/life directly. Most proteins are in 3D structure which gives them an ability to vary their functionality in the life forms. Protein structure and it's structure is exceedingly complex and not easy to predict which makes it very good subject to focus on research.\n",
    "\n",
    "In computer science we represent proteins as 20 different letters putting to gether as a sequence, where each letter represents different amino acids. The amino acids that are linked into a protein chain are often referred to as residues. You will see the term residue used when one wants to refer to a particular amino acid in a particular position of a protein chain."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "### Transcription\n",
    "\n",
    "The process of reading DNA and creating RNA from it is called _transcription_. In computation side of it we will only use the same DNA sequence and will replace DNA nucleotide \"T\" into \"U'. `dna.replace(\"T\", \"U\")` does the job in `Python`.\n",
    "\n",
    "- The DNA strand which has the same sequence as the RNA is called the __coding strand__.\n",
    "\n",
    "### Translation\n",
    "\n",
    "Most RNA molecules go on to specify protein amino acid sequences in a process called __translation__; these are called messenger RNAs (mRNA).\n",
    "\n",
    "- Each subsequent group of _three_ bases, called __codon__.\n",
    "- The regions of an RNA chain that are removed are called __introns__, and those remaining are called __exons.__\n",
    "- Introns' presence makes it significantly more difficult to detect which bits of a gene are actually used to make protein sequences.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Using biological sequences in computing\n",
    "\n",
    "### Important tips\n",
    "\n",
    "- Fit the sequences into some kind of data structures and make sure it will be reusable.\n",
    "- The Commonest and the simplest is storing your sequence as text\n",
    "- DNA and RNA will be represented as strings combined with 4 different characters (A, C, G, T/U)\n",
    "- Proteins will be represented as strings combined with 20 different characters \n",
    "\n",
    "\n",
    "### 1. Translating a DNA sequence into protein\n",
    "\n",
    "In this example, we will translate the given DNA sequence into the protein according to pre-defined structures for aa-protein representation. Right now, we won't think of starting codon, stop codon, or special codons, but we will only implement the main concept of the translation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Here is the genetic code of the amino acids defined as dictionaries\n",
    "STANDARD_GENETIC_CODE = {\n",
    "    \"UUU\": \"Phe\",\n",
    "    \"UUC\": \"Phe\",\n",
    "    \"UCU\": \"Ser\",\n",
    "    \"UCC\": \"Ser\",\n",
    "    \"UAU\": \"Tyr\",\n",
    "    \"UAC\": \"Tyr\",\n",
    "    \"UGU\": \"Cys\",\n",
    "    \"UGC\": \"Cys\",\n",
    "    \"UUA\": \"Leu\",\n",
    "    \"UCA\": \"Ser\",\n",
    "    \"UAA\": None,\n",
    "    \"UGA\": None,\n",
    "    \"UUG\": \"Leu\",\n",
    "    \"UCG\": \"Ser\",\n",
    "    \"UAG\": None,\n",
    "    \"UGG\": \"Trp\",\n",
    "    \"CUU\": \"Leu\",\n",
    "    \"CUC\": \"Leu\",\n",
    "    \"CCU\": \"Pro\",\n",
    "    \"CCC\": \"Pro\",\n",
    "    \"CAU\": \"His\",\n",
    "    \"CAC\": \"His\",\n",
    "    \"CGU\": \"Arg\",\n",
    "    \"CGC\": \"Arg\",\n",
    "    \"CUA\": \"Leu\",\n",
    "    \"CUG\": \"Leu\",\n",
    "    \"CCA\": \"Pro\",\n",
    "    \"CCG\": \"Pro\",\n",
    "    \"CAA\": \"Gln\",\n",
    "    \"CAG\": \"Gln\",\n",
    "    \"CGA\": \"Arg\",\n",
    "    \"CGG\": \"Arg\",\n",
    "    \"AUU\": \"Ile\",\n",
    "    \"AUC\": \"Ile\",\n",
    "    \"ACU\": \"Thr\",\n",
    "    \"ACC\": \"Thr\",\n",
    "    \"AAU\": \"Asn\",\n",
    "    \"AAC\": \"Asn\",\n",
    "    \"AGU\": \"Ser\",\n",
    "    \"AGC\": \"Ser\",\n",
    "    \"AUA\": \"Ile\",\n",
    "    \"ACA\": \"Thr\",\n",
    "    \"AAA\": \"Lys\",\n",
    "    \"AGA\": \"Arg\",\n",
    "    \"AUG\": \"Met\",\n",
    "    \"ACG\": \"Thr\",\n",
    "    \"AAG\": \"Lys\",\n",
    "    \"AGG\": \"Arg\",\n",
    "    \"GUU\": \"Val\",\n",
    "    \"GUC\": \"Val\",\n",
    "    \"GCU\": \"Ala\",\n",
    "    \"GCC\": \"Ala\",\n",
    "    \"GAU\": \"Asp\",\n",
    "    \"GAC\": \"Asp\",\n",
    "    \"GGU\": \"Gly\",\n",
    "    \"GGC\": \"Gly\",\n",
    "    \"GUA\": \"Val\",\n",
    "    \"GUG\": \"Val\",\n",
    "    \"GCA\": \"Ala\",\n",
    "    \"GCG\": \"Ala\",\n",
    "    \"GAA\": \"Glu\",\n",
    "    \"GAG\": \"Glu\",\n",
    "    \"GGA\": \"Gly\",\n",
    "    \"GGG\": \"Gly\",\n",
    "}\n",
    "\n",
    "# Pre-defined DNA sequence, We will use this along the way.\n",
    "dna_seq = \"ATGGTGCATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGGCAAGGTG\"\n",
    "print(\"Input DNA: \")\n",
    "print(dna_seq)\n",
    "\n",
    "print(\"Output RNA:\")\n",
    "# You can also directly change DNA to RNA and save it as rna_seq\n",
    "rna_seq = dna_seq.replace(\"T\", \"U\")\n",
    "print(rna_seq)\n",
    "\n",
    "\n",
    "def protein_translation(seq, genetic_code):\n",
    "    \"\"\"Translate a nucleic acid sequence into a protein sequence.\n",
    "\n",
    "    Translation proceeds until the end of the sequence or until a stop codon is reached.\n",
    "\n",
    "    Args:\n",
    "        seq (str): The nucleic acid sequence to translate.\n",
    "        genetic_code (dict): The genetic code dictionary.\n",
    "\n",
    "    Returns:\n",
    "        list: The translated protein sequence as a list of amino acids.\n",
    "\n",
    "    \"\"\"\n",
    "    # Changes all the T into U, DNA to RNA\n",
    "    seq = seq.replace(\"T\", \"U\")  # Make sure we have RNA sequence\n",
    "    protein_seq = []  # Initializing the protein_seq list to store the output\n",
    "\n",
    "    i = 0\n",
    "    while i + 2 < len(seq):\n",
    "        # Get codons of three letters\n",
    "        codon = seq[i : i + 3]\n",
    "        # Get the match-up amino acid\n",
    "        amino_acid = genetic_code[codon]\n",
    "        # If found stop looping\n",
    "        if amino_acid is None:  # Found stop codon\n",
    "            break\n",
    "        # Other wise add that amino acid to protein_seq list\n",
    "        protein_seq.append(amino_acid)\n",
    "        i += 3\n",
    "\n",
    "    return protein_seq\n",
    "\n",
    "\n",
    "print(\"Output Protein:\")\n",
    "print(protein_translation(dna_seq, STANDARD_GENETIC_CODE))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. Estimating the Molecular Weight\n",
    "\n",
    "This example estimates the weight of a DNA, RNA or protein molecule (in units of Daltons). This is only an estimate because various residues reversibly bind hydrogen ions under different conditions (i.e. pH affects whether ```H+``` ions are joined to the acidic and basic sites) and we are assuming standard proportions of the various isotopes.\n",
    "\n",
    "#### Steps\n",
    "\n",
    "- Define a function with 2 arguments, sequence, and MoleculeType \n",
    "- Define a dictionary inside the function that stores the average molecular weights of the different kinds of residues\n",
    "- Define a variable to hold the total for the molecular mass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define function with seq and mol_type Protein\n",
    "def estimate_mol_mass(seq, mol_type=\"protein\"):\n",
    "    \"\"\"Calculate the molecular weight of a biological sequence.\n",
    "\n",
    "    Assumes normal isotopic ratios and protonation/modification states.\n",
    "\n",
    "    Args:\n",
    "        seq (str): The biological sequence to calculate the molecular weight for.\n",
    "        mol_type (str, optional): The type of molecule, either \"DNA\", \"RNA\", or\n",
    "          \"protein\".  Defaults to \"protein\".\n",
    "\n",
    "    Returns:\n",
    "        float: The molecular weight of the biological sequence in daltons.\n",
    "\n",
    "    \"\"\"\n",
    "    # Define a function with Molecule Masses\n",
    "    residue_masses = {\n",
    "        \"DNA\": {\"G\": 329.21, \"C\": 289.18, \"A\": 323.21, \"T\": 304.19},\n",
    "        \"RNA\": {\"G\": 345.21, \"C\": 305.18, \"A\": 329.21, \"U\": 302.16},\n",
    "        \"protein\": {\n",
    "            \"A\": 71.07,\n",
    "            \"R\": 156.18,\n",
    "            \"N\": 114.08,\n",
    "            \"D\": 115.08,\n",
    "            \"C\": 103.10,\n",
    "            \"Q\": 128.13,\n",
    "            \"E\": 129.11,\n",
    "            \"G\": 57.05,\n",
    "            \"H\": 137.14,\n",
    "            \"I\": 113.15,\n",
    "            \"L\": 113.15,\n",
    "            \"K\": 128.17,\n",
    "            \"M\": 131.19,\n",
    "            \"F\": 147.17,\n",
    "            \"P\": 97.11,\n",
    "            \"S\": 87.07,\n",
    "            \"T\": 101.10,\n",
    "            \"W\": 186.20,\n",
    "            \"Y\": 163.17,\n",
    "            \"V\": 99.13,\n",
    "        },\n",
    "    }\n",
    "\n",
    "    # Get the mol_type from the dictionary residue_masses\n",
    "    mass_dict = residue_masses[mol_type]\n",
    "    # Begin with mass of extra end atoms H + OH\n",
    "    mol_mass = 18.02\n",
    "\n",
    "    # Loop through each letter in sequence\n",
    "    for letter in seq:\n",
    "        # Add the molecule mass according to match-up and sum them up\n",
    "        mol_mass += mass_dict.get(letter, 0.0)\n",
    "\n",
    "    # Return to mol_mass\n",
    "    return mol_mass\n",
    "\n",
    "\n",
    "# Test Case 1\n",
    "print(\"DNA: \" + dna_seq)\n",
    "# Function called with dna_seq variable and DNA mol_type\n",
    "print(\"Weight: \", estimate_mol_mass(dna_seq, mol_type=\"DNA\"), \" daltons\")\n",
    "\n",
    "# Test Case 2\n",
    "print(\"RNA: \" + rna_seq)\n",
    "# Function called with dna_seq variable and RNA mol_type\n",
    "print(\"Weight: \", estimate_mol_mass(rna_seq, mol_type=\"RNA\"), \" daltons\")\n",
    "\n",
    "# Test Case 3\n",
    "# Protein Sequence is defined\n",
    "protein_seq = \"IRTNGTHMQPLLKLMKFQKFLLELFTLQKRKPEKGYNLPIISLNQ\"\n",
    "print(\"Protein: \" + protein_seq)\n",
    "# Function called with protein_seq variable and default value protein\n",
    "print(\"Weight: \", estimate_mol_mass(protein_seq), \" daltons\")\n",
    "print()"
   ]
  }
 ],
 "metadata": {},
 "nbformat": 4,
 "nbformat_minor": 0
}
